qchem/tests/test_deriv_grad.py

```diff
@@ -0,0 +1,96 @@
+import numpy as np
+import pytest
+
+import pennylane as qml
+from pennylane import qchem
+
+import shutil
+
+
+non_zero_coeffs = np.array(
+    [
+        -0.7763135743293005,
+        -0.08534360840293387,
+        -0.08534360840293387,
+        0.2669341092545041,
+        0.26693410925450134,
+        -0.025233628744274508,
+        0.0072162443961340415,
+        -0.0072162443961340415,
+        -0.0072162443961340415,
+        0.0072162443961340415,
+        -0.030654287745411964,
+        -0.023438043349280003,
+        -0.023438043349280003,
+        -0.030654287745411964,
+        -0.02494407786332001,
+    ]
+)
+non_zero_ops = [
+    qml.Identity(wires=[0]),
+    qml.PauliZ(wires=[0]),
+    qml.PauliZ(wires=[1]),
+    qml.PauliZ(wires=[2]),
+    qml.PauliZ(wires=[3]),
+    qml.PauliZ(wires=[0]) @ qml.PauliZ(wires=[1]),
+    qml.PauliY(wires=[0]) @ qml.PauliX(wires=[1]) @ qml.PauliX(wires=[2]) @ qml.PauliY(wires=[3]),
+    qml.PauliY(wires=[0]) @ qml.PauliY(wires=[1]) @ qml.PauliX(wires=[2]) @ qml.PauliX(wires=[3]),
+    qml.PauliX(wires=[0]) @ qml.PauliX(wires=[1]) @ qml.PauliY(wires=[2]) @ qml.PauliY(wires=[3]),
+    qml.PauliX(wires=[0]) @ qml.PauliY(wires=[1]) @ qml.PauliY(wires=[2]) @ qml.PauliX(wires=[3]),
+    qml.PauliZ(wires=[0]) @ qml.PauliZ(wires=[2]),
+    qml.PauliZ(wires=[0]) @ qml.PauliZ(wires=[3]),
+    qml.PauliZ(wires=[1]) @ qml.PauliZ(wires=[2]),
+    qml.PauliZ(wires=[1]) @ qml.PauliZ(wires=[3]),
+    qml.PauliZ(wires=[2]) @ qml.PauliZ(wires=[3]),
+]
+
+
+x = np.array([0.0, 0.0, 0.35, 0.0, 0.0, -0.35])
+
+
+def H(x):
+    return qchem.molecular_hamiltonian(["H", "H"], x)[0]
+
+
+@pytest.mark.parametrize(
+    ("i", "coeffs", "ops"),
+    [(0, [], []), (2, non_zero_coeffs, non_zero_ops), (5, -non_zero_coeffs, non_zero_ops)],
+)
+def test_deriv(i, coeffs, ops):
+    r"""Tests the correctness of the nuclear derivative of the electronic Hamiltonian
+    computed by the ``'derivative'`` function. 
+    """
+
+    deriv = qchem.derivative(H, x, i)
+
+    shutil.rmtree("pyscf")
+
+    exp_obs = qml.vqe.Hamiltonian(coeffs, ops)
+
+    assert exp_obs.compare(deriv)
+
+
+def test_grad(tol):
+    r"""Tests function `'gradient'` computing the gradient of the electronic Hamiltonian H(x)."""
+
+    dev = qml.device("default.qubit", wires=4)
+
+    def circuit(param, wires):
+        qml.BasisState(np.array([1, 1, 0, 0]), wires=[0, 1, 2, 3])
+        qml.RY(param, wires=2)
+        qml.CNOT(wires=[2, 3])
+        qml.CNOT(wires=[2, 0])
+        qml.CNOT(wires=[3, 1])
+
+    grad_h = qchem.gradient(H, x)
+
+    shutil.rmtree("pyscf")
+
+    assert len(grad_h) == x.size
+
+    param = 6.07230111451844
+    exp_grad_E = np.array([0.0, 0.0, -0.03540190344912991, 0.0, 0.0, 0.03540190344912991])
+
+    calc_grad_E = np.array([qml.ExpvalCost(circuit, grad, dev)(param) for grad in grad_h])
+
+    assert np.allclose(exp_grad_E, calc_grad_E, **tol)
```

Comment:
I think that we need more thorough testing:
We should test that the correct quantities are obtained for different Hamiltonians, different deltas, and different coordinates

I also wonder if we should raise exception for incorrect inputs. In particular, we should probably raise an error if the coordinates don't have the right format and if the Hamiltonian is not a properly defined callable function (this may be tricky)

Reply:
> I think that we need more thorough testing:
We should test that the correct quantities are obtained for different Hamiltonians, different deltas, and different coordinates

 I will be adding more general tests by gong beyond H2 which might be too specific. Will add test for different deltas. At the same time we have to keep qchem test suite below 1 min.

> I also wonder if we should raise exception for incorrect inputs. In particular, we should probably raise an error if the coordinates don't have the right format

We can test that but ``molecular_hamiltonian`` will rise an error via the ``meanfield`` function is the ``x`` does not conform to the right size. do you mean to check that ``x`` is a Numpy array?. If I am not mistaken, when we refactor qchem and also some of the templates we were advised to remove type-checks for arguments whose type clearly defined in the docstring.

> if the Hamiltonian is not a properly defined callable function (this may be tricky)

I guess for this we want to check that ``H(x)`` is a callable Hamiltonian object?. I have never done it before. are there examples within PL from which I can learn how to do this?.